ivshmem-plain sample: enable ARM64 support

cataldor opened the pull request:
Hello everyone,
the purpose of this patch series is to add support for ARM64 platforms (qemu-based) for the ivshmem-plain sample.

This series is a prelude to a new version of https://github.com/zephyrproject-rtos/zephyr/pull/52283, where a sample for ITS was asked. Before we can produce an ivshmem-doorbell sample (which uses ITS), we would like to add support for the -plain version.

The series is organized in several commits to make the review process easier,  here is the summary of each one:

1. https://github.com/zephyrproject-rtos/zephyr/commit/adbc7a78b29ac0a630d5d1aa30c9a56859b22ef1: Remove a dependency to `soc.h` for `drivers/virtualization/virt_ivshmem.c` which is not required, as nothing this code uses depends on this header. No functional change intended.

2. https://github.com/zephyrproject-rtos/zephyr/commit/ca5462dec4a1464f47f7997e5a638d11160509e1: commit https://github.com/zephyrproject-rtos/zephyr/commit/a96016d74703da0b0858f6278590792f646629d9 removed the possibility of late lookup for ivshmem PCI device. Therefore, early lookup must work. For ARM64, we need the PCIe controller for PCIe support. In practical terms, PCIe controller needs to be initialized before PCIe.

3. https://github.com/zephyrproject-rtos/zephyr/commit/5d83a8effc6c0aa5f9bd16d6224c9f3978e75ab0: add PCIe controller DTS information for qemu-based ARM64 platforms (extracted from QEMU dtb).

4. https://github.com/zephyrproject-rtos/zephyr/commit/450b97fa7b9af221c136e1ff26ee5c7cc92eeadc: move ivshmem-plain demo to another folder. This is to prepare for a future ivshmem-doorbell sample. No functional change intended.

5. https://github.com/zephyrproject-rtos/zephyr/commit/097f80011aaa9a663f4921d97e6e5221404e0b8f: add qemu-based ARM64 specific files for ivshmem-plain demo. No changes to the sample code.

6. https://github.com/zephyrproject-rtos/zephyr/commit/fa9f6d835c51f08d2a9ecb57f178b20589f05fe9: A fix for the recent commit https://github.com/zephyrproject-rtos/zephyr/commit/6cb5c5519e4d8276fd2bca4b183a65a8af3e3526. If ASSERT is enabled, we need the definition of `INT_MAX`

7. https://github.com/zephyrproject-rtos/zephyr/commit/cbc38bcd6e3dd5dd6bf67939e3fbf45d7355d9da: add kvm ARM64 specific files for ivshmem-plain demo. No changes to the sample code.

-------------
NEW COMMITS POST-REVISION

1. https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/95d55f1d239a69323210b06a7d248e9ef664e328: only include msi.h header when doorbell configuration is chosen

2. https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/cd22952660223b1d233ff9f0717e5396c2b60255: remove MSI configuration options from the project file. This sample does not use MSI at all.

-------------
NOTES

1. commit https://github.com/zephyrproject-rtos/zephyr/commit/fa9f6d835c51f08d2a9ecb57f178b20589f05fe9 can be separated into another merge request, but this MR also adds a KVM test for the twister, so we thought it was best to add here since it can be tested (but see next note). If the community prefers, we can remove commit 6 from this MR.

3. At least locally, KVM test only run when I added the tags `simulation` and `default` to the YAML board file for KVM. The discussion in https://github.com/zephyrproject-rtos/zephyr/pull/51932 suggest the test from commit https://github.com/zephyrproject-rtos/zephyr/commit/cbc38bcd6e3dd5dd6bf67939e3fbf45d7355d9da will not run on GitHub. We added the test anyway, in case this situation changes in the future.

Commit:
drivers: ivshmem: remove unnecessary platform dependency

The ivshmem code does not use any platform-specific code; therefore,
remove the dependency to the soc interface.

No functional change intended.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
drivers: pcie: late initialization of pcie when PCIe controller is used

For architectures that rely on a PCIe controller (for example, ARM64),
scanning the PCI space will only succeed after the controller has
initialized. Therefore, in the presence of PCIe controller, the PCIe
initialization is bumped to the next system init level.

In the past, drivers like ivshmem would do a late scan of the PCI space
in case the early scan failed; however, the cited commit removed this
feature and ivshmem fails for ARM64. This commit fix this by making the
early scan succeed.

Fixes: a96016d74703 ("drivers: ivshmem: Remove unnecessary BDF lookup ...")
Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
dts: qemu: a53: kvm: add DTS support for PCIe controller

Add DTS information for qemu-virt-a53 and qemu-kvm-arm64 for PCIe
controller support. Three new bindings are required for the PCIe
controller in ECAM mode.

The DTS information was extracted from QEMU (dumpdtb) with a PCIe device
attached to the virtual machine (ivshmem)

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
tests: drivers: virtualization: move ivshmem test to a new folder

Move the existing ivshmem test to a folder that explicitly identifies
its target: ivshmem-plain. The existing test is not compatible with
ivshmem-doorbell.

ivshmem-doorbell has additional requirements compared to -plain
(ivshmem-server, notification handling), so it is reasonable that
another test is required for it. This commits prepares the way for the new
test.

No funcional change intended.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
tests: drivers: virtualization: enable QEMU ARM64 support for ivshmem-plain

Add board, project, and twister configurations for QEMU ARM64 support for
ivshmem-plain test.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
include: arch: arm64: limits header needed for ASSERT

Cited commit requires the limits.h header to fetch the defintion of
INT_MAX, which is missing today. The build process fails without it.

Fixes: cb5c5519e4 ("arch/arm64: Support runtime frequency")
Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
tests: drivers: virtualization: enable KVM support for ivshmem-plain

Add board, project, and twister configuration for qemu_kvm_arm64 support
on ivshmem-plain test.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
drivers: ivshmem: msi.h header only required for ivshmem-doorbell

Only include the MSI header when ivshmem-doorbell is employed, as the
ivshmem data structures already use the same protection for msi-related
structures.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

Commit:
tests: drivers: virtualization: remove MSI dependency from sample

The ivshmem-plain test does not use MSI at all; remove the dependency
from the project configuration.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>

carlocaione commented on dts/arm64/qemu/qemu-virt-a53.dtsi:
```diff
@@ -91,5 +100,36 @@
 			 */
 			reg = <0x0 0x0 0x0 DT_SIZE_M(64) /* 0x4000000 DT_SIZE_M(64) */>;
 		};
+
+		pcie: pcie@4010000000 {
+			compatible = "pci-host-ecam-generic";
+			device_type = "pci";
+			reg = <0x40 0x10000000 0x00 0x10000000>;
```
Can you update the comment at the beginning of this file with the `dumpdtb` command you used for this?

cataldor replied:
Hello Carlo! thank you for the review.
Actually, the same command in the current comment (i.e., `qemu-system-aarch64 -machine virt -cpu cortex-a53 -nographic -machine dumpdtb=virt.dtb`) already produces the pcie dtb information. So i do not think it needs to be changed.

I'm using host-tools from zephyr sdk 15.1

carlocaione replied:
perfect, thanks.

carlocaione commented on tests/drivers/virtualization/ivshmem/plain/boards/qemu_cortex_a53.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_ARM64_VA_BITS_40=y
+CONFIG_ARM64_PA_BITS_40=y
+
+# CONFIG_PCIE_MSI is not set
+# CONFIG_PCIE_MSI_X is not set
+# CONFIG_PCIE_MSI_MULTI_VECTOR is not set
```
can we remove these?

cataldor replied:
Hello Carlo. 
Yes, we can, but these leads to other dependencies. I believe we should keep this sample like this (i.e., no MSI for ARM64) and enable MSI for a future -doorbell sample we can provide.

If we enable MSI here, we have a dependency for ITS. Then, ITS depends on two other things: first, we need to increase the IRQ size due to LPI support. We also need to provide dynamic memory, because ITS creates tables at runtime (it asked the hardware for for baser page size).

Let's say we enable all these options; this sample (ivshmem-plain) does not use MSI at all, and will not work if you enable ivshmem-doorbell (it requires that `ivshmem_get_vectors` returns zero). This is the intended scenario for ivshmem-plain: just shared memory.

Our proposal is: provide another sample for ivshmem-doorbell and put a comment back in this file saying something like "if you really want to enable MSI here, look at ivshmem-doorbell sample for dependencies related to MSI"

Let me know if you agree or how do you prefer we proceed.

carlocaione replied:
Sorry, probably I was not clear enough. I was not suggesting to enable those but to simply remove those from the conf file entirely since they are commented out. Or if you want to keep those commented out, add a comment explaining why.

cataldor replied:
oooh, yes. I did not get that.
I rewrote history with rebase.. I'm not sure that is the preferred way to address this? If you can, let me know for future reference.

new/changed commits:
https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/95d55f1d239a69323210b06a7d248e9ef664e328: new commit, protected inclusion of `msi.h` header, so that we can disable MSI for this sample

https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/cd22952660223b1d233ff9f0717e5396c2b60255: new commit. Remove MSI configuration options from the sample.

https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/028f88b480a7ae732f6a65a95dbc3f1be35972e3: changed to remove the special `#` lines for `qemu_cortex_a53`

https://github.com/zephyrproject-rtos/zephyr/pull/52823/commits/fe6173fb3e99a4cf2fc69a2c0fda4430118c69fc: changed to remove the special `#` lines for `qemu_kvm_arm64`

carlocaione reviewed (commented):
Just a couple of nits

SgrrZhf commented on dts/arm64/qemu/qemu-virt-a53.dtsi:
```diff
@@ -91,5 +100,36 @@
+				  0x3000000 0x80 0x00        0x80 0x00        0x80 0x00>;
+			#interrupt-cells = <0x01>;
+			interrupt-map-mask = <0x1800 0x00 0x00 0x07>;
+			interrupt-map = <0x00   0x00 0x00 0x01 0x8002 0x00 0x00 0x00 0x03 0x04
```
Could this part be much more readable? For example, replace some numbers with macro. Just a suggestion and I won't block this PR from being merged.

cataldor replied:
Hello SgrrZhf. Thank you for the idea.

I have updated the definition with a new commit. If you could please take a look.

I would like to use `phandle-array` (which seems to me the right choice for the property type); but we could not make it work. So we looked at some other examples (gpio-map), and they were using the `compound` type (`dts/bindings/gpio/gpio-nexus.yaml`). So that is the solution we used here.

The only way we could make it work with `phandle-array` was adding `#` for `interrupt-map`.. but this seems a hack.

I also had to swap some hexadecimal values for decimal so the line would fit in 100 characters (compliance).

Commit:
dts: bindings: pcie: update interrupt-map type to compound

The interrupt-map property specifies both 32-bits values and a phandle;
update the type accordingly.

Update the definition of pcie-host-ecam-generic on qemu arm64 to match
the new type.

Signed-off-by: Rodrigo Cataldo <rodrigo.cataldo@huawei.com>
Co-authored-by: Henri Xavier <datacomos@huawei.com>
